src/main/java/com/conveyal/taui/controllers/AggregationAreaController.java

```diff
@@ -99,47 +119,79 @@ public static AggregationArea createAggregationArea (Request req, Response res)
+            areas.put(maskName, union.union());
+        } else {
+            // Don't union. Name each area by looking up its value for the name property in the request.
+            features.stream().forEach(f -> areas.put(readProperty(f, maskName), (Geometry) f.getDefaultGeometry()));
+        }
+        // 3. Convert to raster grids, then store them. ================================================================
+        areas.forEach((String name, Geometry geometry) -> {
+            Envelope env = geometry.getEnvelopeInternal();
+            Grid maskGrid = new Grid(SeamlessCensusGridExtractor.ZOOM, env.getMaxY(), env.getMaxX(), env.getMinY(), env.getMinX());
+
+            // Store the percentage each cell overlaps the mask, scaled as 0 to 100,000
+            List<Grid.PixelWeight> weights = maskGrid.getPixelWeights(geometry, true);
+            weights.forEach(pixel -> maskGrid.grid[pixel.x][pixel.y] = pixel.weight * 100_000);
+
+            ObjectMetadata metadata = new ObjectMetadata();
+            metadata.setContentEncoding("gzip");
+            metadata.setContentType("application/octet-stream");
+
+            AggregationArea aggregationArea = new AggregationArea();
+            aggregationArea.name = name;
+            aggregationArea.regionId = regionId;
+
+            // Set `createdBy` and `accessGroup`
+            aggregationArea.accessGroup = req.attribute("accessGroup");
+            aggregationArea.createdBy = req.attribute("email");
+
+            try {
+                File gridFile = File.createTempFile(UUID.randomUUID().toString(),"grid");
+                OutputStream os = new GZIPOutputStream(new BufferedOutputStream(new FileOutputStream(gridFile)));
+                maskGrid.write(os);
+                os.close();
+
+                // Create the aggregation area before generating the S3 key so that the `_id` is generated
+                Persistence.aggregationAreas.create(aggregationArea);
+
+                aggregationAreas.add(aggregationArea);
+
+                InputStream is = new BufferedInputStream(new FileInputStream(gridFile));
+                // can't use putObject with File when we have metadata . . .
+                S3Util.s3.putObject(AnalysisServerConfig.gridBucket, aggregationArea.getS3Key(), is, metadata);
+                is.close();
+            } catch (IOException e) {
+                throw new AnalysisServerException("Error processing/uploading aggregation area");
+            }
+
+            tempDir.delete();
+        });
+
+        return aggregationAreas;
+    }
 
-        return aggregationArea;
+    static String readProperty (SimpleFeature feature, String propertyName) {
+        try {
+            return feature.getProperty(propertyName).getValue().toString();
+        } catch (NullPointerException e) {
+            throw new AnalysisServerException("The supplied Name was not a property of the uploaded features. " +
```

Comment:
This method allows the property to be specified as a parameter, but its error message assumes the property is a Name. I would either remove the parameter or parameterize the error message, with a slight preference for the latter.